File: packages/model-gateway/src/model_gateway/rates.py
"""What each model costs, and the evidence that it costs that.

This is the rate table. Until now there were three, and they disagreed: the
Python table that actually prices a call, the dashboard's published card, and
a third that labels a button in the Prompt2Blog UI. The 3.x-to-2.5 sweep left
one of them holding 3.x prices under 2.5 names -- quoting $2.00 per million
for a model that costs $1.25 -- and nothing caught it, because nothing
compared them.

Every rate carries the date it was checked and the page it was checked
against. A rate nobody has verified is a number that looks like evidence, and
an unverified rate that happens to be wrong makes every cost built on it
wrong in the same direction, silently.

Two caveats that outlive any single verification:

* Gemini 2.5 Flash is on introductory pricing **through 2026-12-31**, after
  which input and output both double. This table will silently understate
  every Flash call from 2027-01-01 until someone edits it.
* These are Gemini API list rates. Calls go through Vertex AI, which bills
  under its own SKUs. The two have matched for these models but are not
  guaranteed to.

The numbers themselves are in ``rates.json`` beside this file, because the
dashboard publishes the same table and reads it from TypeScript. One source,
two runtimes. This module is the Python reader and the thing that applies
them.

Rates that nothing calls any more stay in the table, marked ``in_use=False``. Stored
runs were priced with them, and a rate table that forgets is a receipt that
changes after the fact.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

TABLE_PATH = Path(__file__).with_name("rates.json")

GEMINI_API_PRICING = "https://ai.google.dev/gemini-api/docs/pricing"
GOOGLE_MAPS_PRICING = "https://developers.google.com/maps/billing-and-pricing/pricing"

# When every rate below was last checked against its source.
VERIFIED_ON = "2026-09-04"

# Input tokens above which a tiering model switches to its large-context
# rates. Google applies one threshold across the models that tier at all.
LARGE_CONTEXT_THRESHOLD = 200_000


@dataclass(frozen=True)
class ModelRate:
    """USD per million tokens, plus the provenance of those numbers."""

    model: str
    input_per_million: float
    output_per_million: float
    cached_input_per_million: float
    large_input_per_million: float | None = None
    large_output_per_million: float | None = None
    large_cached_input_per_million: float | None = None
    large_context_threshold: int | None = None
    verified_on: str = VERIFIED_ON
    source: str = GEMINI_API_PRICING
    # False for models nothing calls any more, kept to price stored runs.
    in_use: bool = True
    # Set when the rate is known to change on a date.
    note: str | None = None


@dataclass(frozen=True)
class UnpriceableProvider:
    """A provider whose calls are reported with a duration and no cost.

    Listed because absence from the rate table is otherwise indistinguishable
    from an oversight.
    """

    provider: str
    reason: str
    source: str = ""


def _load(path: Path = TABLE_PATH) -> tuple[tuple[ModelRate, ...], tuple["UnpriceableProvider", ...]]:
    """Read the table, refusing a row that cannot be checked.

    Every rate must carry the date it was verified and the page it was
    verified against. A rate nobody has checked is a number that looks like
    evidence, and an unverified rate that happens to be wrong makes every cost
    built on it wrong in the same direction, silently.
    """
    payload = json.loads(path.read_text(encoding="utf-8"))
    loaded: list[ModelRate] = []
    for row in payload.get("models", []):
        model = row.get("model")
        if not isinstance(model, str) or not model:
            raise ValueError(f"{path.name}: a rate has no model name")
        verified = row.get("verifiedOn", "")
        if not re.fullmatch(r"\d{4}-\d{2}-\d{2}", str(verified)):
            raise ValueError(f"{path.name}: {model} has no verification date")
        source = str(row.get("source", ""))
        if not source.startswith("https://"):
            raise ValueError(f"{path.name}: {model} has no source to re-check")
        loaded.append(
            ModelRate(
                model=model,
                input_per_million=float(row["input"]),
                output_per_million=float(row["output"]),
                cached_input_per_million=float(row["cachedInput"]),
                large_input_per_million=row.get("largeInput"),
                large_output_per_million=row.get("largeOutput"),
                large_cached_input_per_million=row.get("largeCachedInput"),
                large_context_threshold=row.get("largeContextThreshold"),
                verified_on=str(verified),
                source=source,
                in_use=bool(row.get("inUse", True)),
                note=row.get("note"),
            )
        )
    if not loaded:
        raise ValueError(f"{path.name} lists no models")

    providers = tuple(
        UnpriceableProvider(
            provider=str(row["provider"]),
            reason=str(row.get("reason", "")),
            source=str(row.get("source", "")),
        )
        for row in payload.get("unpriceable", [])
    )
    return tuple(loaded), providers


_RATES, UNPRICEABLE_PROVIDERS = _load()

MODEL_RATES: dict[str, ModelRate] = {rate.model: rate for rate in _RATES}

UNPRICEABLE_PROVIDER_NAMES = frozenset(
    entry.provider for entry in UNPRICEABLE_PROVIDERS
)


def rate_for(model_name: str) -> ModelRate | None:
    """The rate for a model, or None when the table has never been told."""
    return MODEL_RATES.get(model_name)


def estimated_cost(
    *,
    model_name: str,
    input_tokens: int,
    output_tokens: int,
    cached_input_tokens: int,
) -> float | None:
    """What a call cost, or None when no rate exists for the model.

    None rather than 0.0 on purpose. A zero is a claim that the call was free;
    the absence of a rate is a claim that nobody knows, and those must not
    look the same on a cost chart.
    """
    rate = MODEL_RATES.get(model_name)
    if rate is None:
        return None

    threshold = rate.large_context_threshold
    use_large = threshold is not None and input_tokens > threshold

    input_rate = (
        rate.large_input_per_million
        if use_large and rate.large_input_per_million is not None
        else rate.input_per_million
    )
    output_rate = (
        rate.large_output_per_million
        if use_large and rate.large_output_per_million is not None
        else rate.output_per_million
    )
    cached_input_rate = (
        rate.large_cached_input_per_million
        if use_large and rate.large_cached_input_per_million is not None
        else rate.cached_input_per_million
    )

    uncached_input_tokens = max(0, input_tokens - cached_input_tokens)
    return (
        uncached_input_tokens * input_rate
        + cached_input_tokens * cached_input_rate
        + output_tokens * output_rate
    ) / 1_000_000


def rates_payload() -> dict[str, Any]:
    """The whole table, shaped for the dashboard's Rates tab."""
    return {
        "verifiedOn": VERIFIED_ON,
        "models": [
            {
                "model": rate.model,
                "input": rate.input_per_million,
                "output": rate.output_per_million,
                "cachedInput": rate.cached_input_per_million,
                "largeInput": rate.large_input_per_million,
                "largeOutput": rate.large_output_per_million,
                "largeCachedInput": rate.large_cached_input_per_million,
                "largeContextThreshold": rate.large_context_threshold,
                "verifiedOn": rate.verified_on,
                "source": rate.source,
                "inUse": rate.in_use,
                "note": rate.note,
            }
            for rate in _RATES
        ],
        "unpriceable": [
            {
                "provider": entry.provider,
                "reason": entry.reason,
                "source": entry.source,
            }
            for entry in UNPRICEABLE_PROVIDERS
        ],
        "appliedBy": "packages/model-gateway/src/model_gateway/rates.json",
    }

File: apps/dashboard/src/usage/rates.ts
// There were three copies of the token-rate table: the Python that prices a
// call, the card this file publishes, and a third labelling a button in the
// Prompt2Blog routing panel. The 3.x-to-2.5 sweep left one holding 3.x prices
// under 2.5 names and nothing caught it, because nothing compared them.
// `rates.drift.test.ts` existed to compare this one against that third copy.
// The routing panel was deleted on 2026-09-05 along with the model stacks it
// priced -- v4 has no model picker -- so the third copy is gone and the drift
// test with it. Two remain, and they are the same file: the gateway's
// rates.json, read by the Python and by this card. If a rate table is ever
// written out in a browser bundle again, it needs a test like that one back.
/**
 * What each model costs, and the evidence that it costs that.
 *
 * The collector does not price anything -- reporting apps compute cost and send
 * it. So this is not a pricing engine; it is the published record of the rates
 * those apps are using, next to the date each was checked and the page it was
 * checked against.
 *
 * The numbers are no longer here. They live in the model gateway's
 * `rates.json`, which is also what prices a call, so the card and the code that
 * applies it are the same table rather than two that have to be kept in step.
 * There used to be three copies and they disagreed: the 3.x-to-2.5 sweep left
 * one holding 3.x prices under 2.5 names, quoting $2.00 per million for a model
 * that costs $1.25, and nothing caught it because nothing compared them.
 *
 * Read at runtime rather than imported. The gateway is a Python package outside
 * this app's `rootDir`, and a JSON import would either drag it into the
 * TypeScript build graph or need a copy checked in here -- which is the problem
 * again.
 */

import { readFileSync } from "node:fs";
import { join } from "node:path";
import type { ModelRate, RatesPayload, UnpriceableProvider } from "./rate-types";

export type { ModelRate, RatesPayload, UnpriceableProvider } from "./rate-types";

type RateTableFile = Omit<RatesPayload, "appliedBy">;

export const RATE_TABLE_PATH = join(
  import.meta.dir,
  "../../../../packages/model-gateway/src/model_gateway/rates.json",
);

/**
 * Where the rates are applied, shown on the card so a reader can go and look.
 */
export const APPLIED_BY = "packages/model-gateway/src/model_gateway/rates.json";

function readTable(): RateTableFile {
  // Deliberately unguarded. A missing or malformed rate table is a broken
  // install, not a runtime condition to degrade through, and a card that
  // silently renders nothing is worse than one that fails loudly at boot.
  const parsed = JSON.parse(readFileSync(RATE_TABLE_PATH, "utf8")) as RateTableFile;
  if (!Array.isArray(parsed.models) || parsed.models.length === 0) {
    throw new Error(`${RATE_TABLE_PATH} lists no models`);
  }
  return parsed;
}

let cached: RateTableFile | undefined;

function table(): RateTableFile {
  cached ??= readTable();
  return cached;
}

export function modelRates(): ModelRate[] {
  return table().models;
}

export function unpriceableProviders(): UnpriceableProvider[] {
  return table().unpriceable;
}

export function ratesPayload(): RatesPayload {
  const current = table();
  return {
    verifiedOn: current.verifiedOn,
    models: current.models,
    unpriceable: current.unpriceable,
    appliedBy: APPLIED_BY,
  };
}
